### ISSUE TITLE
K8s 1.23.14 (default) deployment fails

### ISSUE THREAD

**Issue by chess-knight** [SovereignCloudStack/k8s-cluster-api-provider#321]
Similar problem to #303.  
Deployment gets stuck during bootstrap of first control plane because coredns container image cannot be pulled.

**Example log of containerd on control plane:**
```
Jan 25 10:46:36 capi-testcluster-control-plane-genc01-dcck4 containerd[545]: time="2023-01-25T10:46:36.830350539Z" level=info msg="PullImage \"registry.k8s.io/coredns:v1.8.6\""
Jan 25 10:46:37 capi-testcluster-control-plane-genc01-dcck4 containerd[545]: time="2023-01-25T10:46:37.554457011Z" level=error msg="PullImage \"registry.k8s.io/coredns:v1.8.6\" failed" error="rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/coredns:v1.8.6\": failed to resolve reference \"registry.k8s.io/coredns:v1.8.6\": registry.k8s.io/coredns:v1.8.6: not found"
```

I tested it and it seems that *registry.k8s.io* is used only from [v1.23.15](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.23.md#changelog-since-v12314).  
This also applies to [v1.22.17](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#changelog-since-v12216) and [v1.24.9](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md#changelog-since-v1248).  
PR #313 only checks major and minor versions, which seems insufficient.

I found that it was is fixed in cluster-api [v1.2.9](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.9) so maybe we do not need this registry patching logic at all. Maybe we also don't need [`imageRepository`](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/terraform/files/template/cluster-template.yaml#L59) in the KubeadmControlPlane spec. This [issue](https://github.com/kubernetes-sigs/cluster-api/issues/7833) in cluster-api is probably also related.

**Comment by garloff** [comment]
Thanks, @chess-knight !
The registry switch upstream is a PITA.
Trouble with capi >= 1.2.8 and omitting the imageRegistry setting is that certain patch levels will be blocked.
While upgrading to the most recent patch level is by far the most popular choice, I feel like not supporting lower versions is patronizing our users.
So we might need to continue adjusting the imageRegistry setting, but consider the patch levels...
Thoughts?

**Comment by chess-knight** [comment]
I agree we should support lower versions and I think that patching is in this case better option, but if we want to patch the registry location, we should do it correctly. We can use e.g. cluster-api [logic](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/util/kubeadm/kubeadm.go) for that. I also found that this [PR](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/pull/315) adds "critical" kubernetes patch versions as defaults, so at least it will not fail by default as in this case.

**Comment by garloff** [comment]
Yes, I will replicate the logic from capi -- that's the best source of truth indeed.
#315 was not merged b/c v1.26 was included. (It works, but is not officially supported).
I think we can merge #315 now, which makes this one a bit less critical, but of course we'll need to fix it still.

**Comment by garloff** [comment]
Should be fixed by #324.

**Comment by garloff** [comment]
Addressed by merged #324.
